modules/bmc/bmc_api.rb

```diff
@@ -8,6 +8,30 @@ class Api < ::Sinatra::Base
     # All GET requests will only read ipmi data, no changes
     # All PUT requests will update information on the bmc device
 
+    # we have to parse the JSON body in the request and merge into the params object for later use
+    # if the user does not set the content type to application/json we will just assume there is garbage in the body
+    # also if the user decides to do http://127.0.0.1/bmc/192.168.1.6/test?bmc_provider=freeipmi as well as pass in
+    # a json encode body with the parameters, all of these items will be merged together
+    before do
+      content_type :json
+      if request.env["CONTENT_TYPE"] == "application/json"
+        begin
+
+          body_parameters = request.body.read
+          unless body_parameters.nil? || body_parameters.empty?
+            json_data = JSON.parse(body_parameters)
+            # convert all keys to symbols
+            json_data.keys.each do |key|
+              json_data[(key.to_sym rescue key) || key] = json_data.delete(key)
+            end
+            params.merge!(json_data)
```

Comment:
Frankly, I don't like merging anything into params hash at all. This can bring security issues in the future (if we authorize users per host against URL then this is vulnerable to passing a different host accessing arbitrary system).

Why don't you create a helper method that parses the body and returns what you need?

Other than this one :+1: It took me an hour on Friday to realize this actually does not work...